dxp/react_rendering.md

```diff
@@ -0,0 +1,148 @@
+# React Rendering in DXP
+
+This article describes DXP infrastructure to render React components from the server side perspective. This includes taglibs as well as low-level services.
+
+## React Renderer Service
+
+-   OSGi bundle: `com.liferay.portal.template.react.renderer.api`
+-   Interface name: `com.liferay.portal.template.react.renderer.ReactRenderer`
+-   Service type: `singleton`
+
+This is the low-level service to render React components.
+
+> Note that it renders React components, not any React code, so you must provide a source which exports a React component to render.
+
+It has one method:
+
+```java
+public void renderReact(
+		ComponentDescriptor componentDescriptor, Map<String, Object> data,
+		HttpServletRequest httpServletRequest, Writer writer)
+	throws IOException;
+```
+
+which receives:
+
+1. A `ComponentDescriptor` describing the source of the React component to render. More on this in the next section.
+2. A `Map` containing arbitrary data that will be injected to the React component as initial properties. Note that the data inside the map must be compatible with JavaScript, so you cannot inject any Java object you like, only primitive types, `Map`s, `Array`s, ... that can be serialized to JSON.
+3. The current `HttpServletRequest` for which the component is being rendered.
+4. A `Writer` where the rendered HTML (and optionally the JavaScript) is written to.
```

Comment:
Is JS really "optional` here? We always render JS to render the React component, don't we?

Reply:
I will rephrase that. 

I mean that the JS gets written optionally to that Writer in the sense that it is only written if you set `positionInline` to true. Otherwise it is attached to a ScriptData and gets written to the request's Writer at the end of the request's lifecycle.